scripts/dev/generate_config_types.py

```diff
@@ -0,0 +1,237 @@
+    """
+    result = {}
+    for key, value in config.items():
+        parts = key.split(".")
+        current = result
+        for part in parts[:-1]:
+            current = current.setdefault(part, {})
+        current[parts[-1]] = value
+    return result
+
+
+def _get_type_hint(typ: configtypes.ConfigType) -> str:  # noqa: C901
+    if isinstance(typ, configtypes.List):
+        return f"list[{get_type_hint(typ.valtype)}]"
+
+    if isinstance(typ, configtypes.Dict):
+        return f"Mapping[{get_type_hint(typ.keytype)}, {get_type_hint(typ.valtype)}]"
+
+    if isinstance(typ, configtypes.Bool):
+        return "bool"
+
+    valid_values = typ.get_valid_values()
+    if valid_values:
+        return f'Literal[{", ".join(literal(k) for k in valid_values.values)}]'
+    if isinstance(typ, configtypes.MappingType):
+        # note: should be redundant, but means we can get exhaustive type matching
+        return f'Literal[{", ".join(literal(k) for k in typ.MAPPING.keys())}]'
+
+    if isinstance(
+        typ,
+        (
+            configtypes.String,
+            configtypes.FontBase,
+            configtypes.QtColor,
+            configtypes.QssColor,
+            configtypes.Command,
+            configtypes.Key,
+            configtypes.SessionName,
+            configtypes.Url,
+            configtypes.File,
+            configtypes.UrlPattern,
+            configtypes.FormatString,
+            configtypes.Directory,
+            configtypes.Encoding,
+            configtypes.FuzzyUrl,
+            configtypes.SearchEngineUrl,
+            configtypes.Proxy,
+        ),
+    ):
+        return "str"
+
+    if isinstance(typ, configtypes.Int):
+        return "int"
+
+    if isinstance(typ, configtypes.Float):
+        return "float"
+
+    if isinstance(typ, configtypes.ListOrValue):
+        return f"Union[{get_type_hint(typ.valtype)}, {get_type_hint(typ.listtype)}]"
+
+    if isinstance(typ, configtypes.Perc):
+        return "Union[float, int, str]"
+
+    if isinstance(typ, configtypes.PercOrInt):
+        return "Union[int, str]"
+
+    if isinstance(typ, configtypes.Regex):
+        return "Union[str, re.Pattern[str]]"
+
+    assert_never(typ)
+
+
+def get_type_hint(typ: configtypes.BaseType) -> str:
+    """Turn a config type into a Python type annotation, e.g. `String` -> `'str'`."""
+    typ = cast(configtypes.ConfigType, typ)
+
+    inner = _get_type_hint(typ)
+
+    if typ.none_ok:
+        return f"Optional[{inner}]"
+
+    return inner
+
+
+def generate_config_types(config_data: Mapping[str, Option]) -> str:
+    """Generate the `ConfigContainer` dataclass and all nested types."""
+    output = [
+        triple_quote(
+            dedent("""
+                This file defines static types for the `c` variable in `config.py`.
+
+                This is auto-generated from the `scripts/dev/generate_config_types.py` file.
+
+                It is not intended to be used at runtime.
+
+                Example usage:
```

Comment:
Once this is definitive, we should also adjust [`ConfigPyWriter`](https://github.com/qutebrowser/qutebrowser/blob/v3.3.1/qutebrowser/config/configfiles.py#L800-L929) so that it adds such a header in the generated `config.py` file.